src/Language/PureScript/TypeChecker/Kinds.hs

```diff
@@ -409,17 +410,41 @@ checkTypeKind
   -> SourceType
   -> m ()
 checkTypeKind ty kind =
-  unifyKindsWithFailure (\_ _ -> throwError . errorMessage $ ExpectedType ty kind) kind E.kindType
+  -- In this case, if any failure occurs, we ignore it and throw our own error.
+  -- So, rather than use `unifyishRows` here and pay for its overhead,
+  -- we'll use the approach that was used before we migrated to `unifyishRows`
+  -- because the original approach fails faster.
+  unifyKindsWithFailure (UseSequential ty kind) kind E.kindType
+
+data UnifyRowOptions m
+  -- type kind
+  = UseSequential SourceType SourceType
+  -- isExpectedError onFailure
+  | UseUnifyishRows (SimpleErrorMessage -> Bool) (SourceType -> SourceType -> m MultipleErrors)
 
 unifyKindsWithFailure
   :: (MonadError MultipleErrors m, MonadState CheckState m, HasCallStack)
-  => (SourceType -> SourceType -> m ())
+  => UnifyRowOptions m
   -> SourceType
   -> SourceType
   -> m ()
-unifyKindsWithFailure onFailure = go
+unifyKindsWithFailure opts = go
   where
-  goWithLabel l t1 t2 = withErrorMessageHint (ErrorInRowLabel l) $ go t1 t2
+  (unifyRows, onFailure) = case opts of
+    UseSequential ty kind ->
+      ( \r1 r2 -> do
+          let (matches, rest) = alignRowsWith (\_ -> go) r1 r2
+          sequence_ matches
+          void $ unifyTails rest
+
+      , \_ _ -> 
+          pure $ errorMessage $ ExpectedType ty kind
```

Comment:
This is a bit of a leaky abstraction, isn't it? `UseSequential` selects not only which strategy to use but also the type of the error message. Maybe both constructors should take an `onFailure`.

In the other direction, does the `UseUnifyishRows` constructor ever take an `isExpectedError` that isn't `isKindsDoNotUnify`? Maybe that should be pushed down into the implementation here.